src/formatter.rs

```diff
@@ -488,18 +489,20 @@ impl FormatRecorder {
     }
 
     pub fn replay(&self, formatter: &mut dyn Formatter) -> io::Result<()> {
-        self.replay_with(formatter, |formatter, data| formatter.write_all(data))
+        self.replay_with(formatter, |formatter, range| {
+            formatter.write_all(&self.data[range])
+        })
     }
 
     pub fn replay_with(
         &self,
         formatter: &mut dyn Formatter,
-        mut write_data: impl FnMut(&mut dyn Formatter, &[u8]) -> io::Result<()>,
+        mut write_data: impl FnMut(&mut dyn Formatter, Range<usize>) -> io::Result<()>,
```

Comment:
It seems unfortunate that data content now has to be tracked separately from its indices.

One thought I had is that perhaps, `write_data` could receive a formatter (in a state with no labels active), a list of labels, and a sequence of bytes? Alternatively, I know you weren't a fan of FormattedString, but maybe the FormatRecorder could return some variation on `Vec< (Vec<Label>, [u8])>` that the caller could process?

Then, all the byte counting would happen strictly inside the wrapping algorithm and wouldn't be exposed in the API. It would also allow replacing the `data` method with a utility function that calls `replay_with` with a function that ignores labels (not that this is important).

I'm guessing you have some future uses of FormatRecorder in mind that make this approach more desirable?

Reply:
> It seems unfortunate that data content now has to be tracked separately from its indices.
> 
> One thought I had is that perhaps, `write_data` could receive a formatter (in a state with no labels active), a list of labels, and a sequence of bytes? Alternatively, I know you weren't a fan of FormattedString, but maybe the FormatRecorder could return some variation on `Vec< (Vec<Label>, [u8])>` that the caller could process?

Yeah, that's another way to deal with. In that design, word splitting logic would take `[(Vec<Label>, [u8])]`, and split them ignoring label boundary. We would need a special case for `<prefix>dead</prefix><rest>beef</rest>` there because these two chunks should be considered a single word.

I think splitting unlabeled `[u8]` bytes is simpler (and generally useful) than splitting labeled chunks, so I moved the complexity to `write_wrapped()` instead. 

Regarding `FormatRecorder`, it serves the same role as `FormattedString`. The main difference is that the caller decides whether or not buffer the output.

Reply:
The splitting logic could again work in two passes, just like now. First generate the unlabeled [u8] inside the wrapping template, then wrap it with the same function you wrote, and finally go over the `[(Vec<Label>, [u8])]` counting out the bytes. The main difference is that all the messy bits (by which I mean the byte counting and the correspondence between indices and data) is now a private implementation detail of the wrapping template.

Reply:
Doesn't it increase the complexity? We would have to reconstruct offsets from the `[u8]` of `[(Vec<Label>, [u8])]`, which the `FormatRecorder` knows. And the wrapping function isn't interested in labels.

Reply:
If we stick with this approach (if we don't switch to formatting `[(Vec<Label>, [u8])]`), we could make this function private and make `replay_wrapped` a method of FormatReplayer instead of a separate method as you have it now. I think that would make this feel a lot tidier at little cost.

Reply:
`replay_with` as it was before can stay public, of course.

Reply:
Maybe you mean `.replay_with()` for byte slices, and `.replay_with_range()` for ranges?

Word-wrapping isn't a business of FormatRecorder.

Reply:
No, I meant what your second paragraph rejects. 

I'm not sure if it'd be worth it at that point, but if text wrapping is not a business of the FormatRecorder, you could have an abstraction  for transformations `FormatRecorder -> FormatRecorder` that's defined next to FormatRecorder and has access to its internals, separate from the templating related parts of the ReformattingTemplate.

Reply:
I don't think it's good idea. `FormatRecorder` is just a buffer like `Vec<u8>`. Functions to process the buffered content shouldn't belong to the recorder.

Reply:
OK, let's go with what you have then.